lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+
+type eventType int
+
+const (
+	nonRemove eventType = 1
+	remove              = 2
+	mixed               = 3
+)
+
+// Not meant to be changed, but must be changeable for tests
+var (
+	maxFiles       = 512
+	maxFilesPerDir = 128
+)
+
+type event struct {
+	path         string
+	firstModTime time.Time
+	lastModTime  time.Time
+	evType       eventType
+}
+
+type eventBatch map[string]*event
+
+type eventDir struct {
+	path      string
+	parentDir *eventDir
+	events    eventBatch
+	dirs      map[string]*eventDir
+}
+
+func newEventDir(path string, parentDir *eventDir) *eventDir {
+	return &eventDir{
+		path:      path,
+		parentDir: parentDir,
+		events:    make(eventBatch),
+		dirs:      make(map[string]*eventDir),
+	}
+}
+
+type watcher struct {
+	folderID            string
+	folderPath          string
+	folderDescription   string
+	folderIgnorePerms   bool
+	folderIgnores       *ignore.Matcher
+	folderIgnoresUpdate chan *ignore.Matcher
+	// time interval to search for events to be passed to syncthing-core
+	notifyDelay time.Duration
+	// time after which an active event is passed to syncthing-core
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	notifyTimerResetChan  chan time.Duration
+	notifyChan            chan []string
+	// All detected and to be scanned events are stored in a tree like
+	// structure mimicking folders to keep count of events per directory.
+	rootEventDir     *eventDir
+	backendEventChan chan notify.EventInfo
+	inProgress       map[string]struct{}
+	cfg              *config.Wrapper
+	configUpdate     chan config.FolderConfiguration
+	stop             chan struct{}
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	C() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+	VerifyConfiguration(from, to config.Configuration) error
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(id string, cfg *config.Wrapper, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &watcher{
+		folderID:              id,
+		folderIgnores:         ignores,
+		folderIgnoresUpdate:   make(chan *ignore.Matcher),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		backendEventChan:      make(chan notify.EventInfo, maxFiles),
+		inProgress:            make(map[string]struct{}),
+		cfg:                   cfg,
+		configUpdate:          make(chan config.FolderConfiguration),
+		stop:                  make(chan struct{}),
+	}
+	folderCfg, ok := cfg.Folder(id)
+	if !ok {
+		panic(fmt.Sprintf("bug: Folder %s does not exist", id))
+	}
+	fsWatcher.updateConfig(folderCfg)
+
+	if err := fsWatcher.setupBackend(); err != nil {
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (w *watcher) setupBackend() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, w.folderPath) {
+			return true
```

Comment:
I think it shouldn't happen, yes, but it is not in our direct control, as the arguments come from the backend. However I guess it is a good idea, otherwise a problem here could silently drop all changes.